pyiron_contrib/workflow/node.py

```diff
@@ -0,0 +1,252 @@
+        function, but it is also possible to pass in any class instance where `__call__`
+        is defined. These attributes are intended to operate in a purely functional way
+        (i.e. without internal state), but it's possible a class-based description may
+        be useful.
+
+    Warning:
+        `input_channels` and `output_channels` are mutable class-level variables.
+        Modifying them in any way is probably unwise.
+
+    Examples:
+        Instantiating from `Node`:
+        >>> from pyiron_contrib.workflow.node import Node
+        >>> from pyiron_contrib.workflow.channels import ChannelTemplate
+        >>>
+        >>> def start_to_end(a=None):
+        ...     return {"x": a}
+        >>>
+        >>> def add_one(x=None):
+        ...    return {"y": x + 1}
+        >>>
+        >>> my_adder = Node(
+        ...     "my_adder",
+        ...     input_channels=[ChannelTemplate("a", types=(int, float))],
+        ...     preprocessor=start_to_end,
+        ...     engine=add_one,
+        ...     # We'll leave the post-processor empty and just align our output
+        ...     # with what our engine returns
+        ...     output_channels=[ChannelTemplate("y")],
+        ... )
+        >>> my_adder.output.y.value
+
+        >>> # Nothing! It tried to update automatically, but there's no default for
+        >>> # "a", so it's not ready!
+        >>> my_adder.input.a.update(1)
+        >>> my_adder.output.y.value
+        2
+
+        Subclassing `Node`:
+        >>> from pyiron_contrib.workflow.node import Node, pass_all
+        >>> from pyiron_contrib.workflow.channels import ChannelTemplate
+        >>>
+        >>> class ThreeToOne(Node):
+        ...     # Expects an engine that maps three numeric values (xyz) to one (w)
+        ...     input_channels = [
+        ...         ChannelTemplate("x", default=1, types=(int, float)),
+        ...         ChannelTemplate("y", default=2, types=(int, float)),
+        ...         ChannelTemplate("z", default=3, types=(int, float)),
+        ...     ]
+        ...     preprocessor = staticmethod(pass_all)
+        ...
+        ...     @staticmethod
+        ...     def postprocessor(**kwargs):
+        ...         return pass_all(**kwargs)
+        ...     # Neither pre- nor post-processor does anything here,
+        ...     # they're just exampels of how to declare them as static
+        ...
+        ...     output_channels = [
+        ...         ChannelTemplate("w", types=(int, float)),
+        ...     ]
+        ...
+        ...     def __init__(self, name: str, engine: callable, **kwargs):
+        ...         # We'll modify what's available in init to push our users a certain direction.
+        ...         super().__init__(name=name, engine=engine, **kwargs)
+        >>>
+        >>> def add(x, y, z):
+        ...     return {"w": x + y + z}
+        >>>
+        >>> adder = ThreeToOne("add", add)
```

Comment:
I would say the hurdle should be even lower for new users (i.e. if I have `def add`, I don't want to write the full class in addition). Here's my suggestion: We ask the user to write the function in the format:

```python
def add(x:int | float = 1, y:int | float = 2, z:int | float = 3) -> int | float:
    return x + y + z
```

Then we should be able to do

```python
adder = Node("add", add)
```

And internally you can do something like this:

```python
import inspect

new_node = Node()
for key, value in inspect.signature(add).parameters.items():
    new_input = {'name': key}
    if value.annotation is not inspect.Parameter.empty:
        new_input['types'] = value.annotation
    if value.default is not inspect.Parameter.empty:
        new_input['default'] = value.default
    new_node.input_channels.append(ChannelTemplate(**new_input))
```

Here, I didn't write the output channel, because I was not really sure how to cope with the fact that there's no name, but at least the type can be obtained by `inspect.signature(add).return_annotation`.

Reply:
Note: The last cell should be rewritten when it's included inside `class Node`. I wrote it in this format so that it works in the notebook.

Reply:
@liamhuber ?

Reply:
For input I like it, but it is problematic for output. Output is a dict, so even the type doesn't help us much. We might be able to nest types in the dict like like lost or tulle (I haven't tested this), but we're still in trouble because the dictionary is not ordered.

Ultimately I care more about the syntax for input and output being the same than for the syntactic sugar of not needing to specify channels, so if we can't figure a way to get the output running with this, I wouldn't really want it for input alone.

Reply:
Why does output have to be a dict?

Reply:
I have to say it makes a huge difference whether the user has to write this

```python
from pyiron_contrib.workflow.node import Node
from pyiron_contrib.workflow.channels import ChannelTemplate

class ThreeToOne(Node):
    input_channels = [
        ChannelTemplate("x", default=1, types=(int, float)),
        ChannelTemplate("y", default=2, types=(int, float)),
        ChannelTemplate("z", default=3, types=(int, float)),
    ]
    output_channels = [
        ChannelTemplate("w", types=(int, float)),
    ]
    
def add(x, y, z):
    return {"w": x + y + z}

adder = ThreeToOne("add", add)
```

or this

```python
from pyiron_contrib.workflow.node import Node

def add(x:int | float = 1, y:int | float = 2, z:int | float = 3) -> int | float:
    return x + y + z

adder = Node("add", add)
```

It's just not realistic to force people to write the upper variant.

Reply:
You are right, it doesn't _have_ to be, but it's for basically the same rationale that the input is kwargs. If you are using just args as input and returning a tuple as output, then at the graph level you're writing connections like this `downstream_node.inputs[2] = upstream_node.outputs[2]`. 

This is actually how ryvencore is set up, and I can tell you from personal experience that it's a huge pain. Without names you lose track of what is going on so quickly. One of the very early things we did in ironflow was overload the ryven functionality to add dot-based access using the channel (port) names, and hit was a huge QoL improvement.

Similarly, if you think about a GUI representation of the graph, on the output side it just gets these boxes `output0; output1; output2`, so you have no immediate idea where they should go. You might show the output type, but there will absolutely be times where a function returns two types and you wouldn't know which is which without more carefully inspecting the innards of the node function (e.g. think about returning energies and volumes -- they'd both just be `np.array[float]`).

Actually, the more we talk about it the more convinced I am that forcing node developers to explicitly declare their channels with names is the right way to go. It's a bit more verbose, but rather than being a "hurdle" I think it's a help by making everything clear and explicit.

I took a quick look through the inspect docs, and I don't see any way to pull out the keys of a returned dictionary systematically. We could get the string source code and manually parse the string for stuff after `return`, but this seems _super_ fragile, e.g. if they define the dict early and then just `return my_dict`. Being robust would require a bunch of crazy code analysis.

Reply:
> It's just not realistic to force people to write the upper variant.

I disagree. Also, `inspect` for sure won't handle anything with ontological typing, e.g. `ChannelTemplate('my_typed_thing', types=Atoms, otype=onto.AtomicStructure).

IMO there's nothing _hard_ or _complicated_ about the top example, it's just a bit verbose. People still write in C, so obviously there's plenty of folks willing to put up with some verbosity.

If you can find a way to get the bottom example working without throwing away the naming of channels at the graph level (i.e. output), I'm open to it. Although we'll need to be careful that the complexity we add by including all the implicit magic is clear enough to developers.

Reply:
>  I don't want to write the full class in addition

Also you don't have to. Here is your lower example sans class. Definitely still more verbose, but keeps the output channel name.

```
from pyiron_contrib.workflow.node import Node
from pyiron_contrib.workflow.channels import ChannelTemplate
# In the long run we'll obviously give a single point of import for everything

def add(x:int | float = 1, y:int | float = 2, z:int | float = 3) -> dict:
    return {"w": x + y + z}

adder = Node(
    "add", 
    input_channels = [
        ChannelTemplate("x", types=(int, float)),
        ChannelTemplate("y", types=(int, float)),
        ChannelTemplate("z", types=(int, float)),
    ],
    engine=add,
    output_channels [ChannelTemplate("w", types=(int, float))]
)
```

Reply:
> Also, `inspect` for sure won't handle anything with ontological typing, e.g. `ChannelTemplate('my_typed_thing', types=Atoms, otype=onto.AtomicStructure).

This argument is not valid, because not only we don't have it and also not in the example, but also even when we have it, we will most likely not make it mandatory. So the fact that the function definition cannot have the ontological typing does not make it less valuable. Besides there's definitely a way to define it as well:

```python
factory = Node("factory", create_atoms)
factory.my_typed_thing.otype = onto.AtomicStructure
```

> IMO there's nothing _hard_ or _complicated_ about the top example, it's just a bit verbose. People still write in C, so obviously there's plenty of folks willing to put up with some verbosity.

I disagree. In contrast to classes like `AtomicGenericJob`, which just have to be useful to developers, the workflow class must be easy to use for anyone making workflows, i.e. virtually everyone, if not literally everyone.

After having written this last phrase, I started realising that maybe this is where the difference between you and me comes from. For me, this new workflow class is on the one hand going to be used by developers to replace mainly the current master classes (in which case, yes I agree that they **must** explicitly say what are the output values), but also by anyone writing a code in the notebook, e.g. master students, early phd students or experimentalists who don't really have a clear idea of how to write a proper code, or also for advanced users who want to have a lot of small nodes quickly. In this case, it is important that a simple function can be included in the workflow in terms of pyiron.

Reply:
Maybe we can also involve @pmrv and @srmnitc because they probably also have good insight into this.

Reply:
> This argument is not valid, because not only we don't have it and also not in the example, but also even when we have it, we will most likely not make it mandatory. 

You're 100% write, I was being too hasty and am wrong. We [_do_ have it](https://github.com/pyiron/ironflow/blob/1e5195cdc68d80bd9256a002363baf8724addb72/ironflow/model/port.py#L56), but not having it in the "function" approach doesn't preclude having it in the OOP approach, so you'll get no further argument from me.

> After having written this last phrase, I started realising that maybe this is where the difference between you and me comes from. For me, this new workflow class is on the one hand going to be used by developers to replace mainly the current master classes (in which case, yes I agree that they must explicitly say what are the output values), but also by anyone writing a code in the notebook, e.g. master students, early phd students or experimentalists who don't really have a clear idea of how to write a proper code, or also for advanced users who want to have a lot of small nodes quickly. In this case, it is important that a simple function can be included in the workflow in terms of pyiron.

I don't think we have conflicting visions about who will use it, but maybe pretty different perspectives on how they should be treated. The pythonic experiment of treating everyone like an adult is a couple decades old, and has led to python being the language of choice for scientists and casuals, but IMO it's also a key contributor to so much scientific code being a hot pile of **** -- not even just at the architecture level, but even at the level of individual functions. It is precisely people "who don't really have a clear idea of how to write a proper code" who I have in mind when I make a framework where, at a minimum, the input and output of each piece of code need to be _named_. The tricky bit is that for a workflow paradigm, it's important to have named outputs as well as named inputs. That's perfectly in line with OOP, where you can populate class attributes, but the only way to accomplished named output at the function level in python is by returning a dict. 

Your example was shorter precisely because it dropped this piece of information. I agree that we don't want to force everyone to make subclasses, but we can recover it without forcing people to subclass `Node` as shown in my example above. I agree that it's a little bit verbose, but I don't think instantiating a second class (`ChannelTemplate`) is particularly difficult, and because of my perspective from the previous paragraph I am willing to pay some verbosity if in return we can force people to at least take a moment to name things.

That said, I don't have any objection in principal to your example of getting it all done in a single function call _if_ we can preserve (a) multiple output values, and (b) those output fields being explicitly named, and (c) the ability to associate types with the output fields (although I could be flexible on the last point). Ideally we could write something like 

```python
from pyiron_contrib.workflow.node import Node

def add(x:int | float = 1, y:int | float = 2, z:int | float = 3) -> dict["w": int | float]:
    return x + y + z

adder = Node("add", add)
```

But python just doesn't support this sort of behaviour, and I don't currently see any implementation that gets the job done.

Reply:
> But python just doesn't support this sort of behaviour, and I don't currently see any implementation that gets the job done.

Actually, it looks like we can do it by type hinting [a TypedDict](https://peps.python.org/pep-0589/), but as far as I can see that involves first defining a TypedDict -- like, at the notebook level for your example, which I think kind of defeats the purpose and is just more verbose than defining a set of `output_channels`.

Reply:
Look, this is a [Dask example they give on their website](https://examples.dask.org/delayed.html):

```python
def inc(x):
    return x + 1

def dec(x):
    return x - 1

def add(x, y):
    return x + y
```

And then they can be executed like this:

```python
x = inc(1)
y = dec(2)
z = add(x, y)
z
```

The first cell of this example, with the current implementation looks like this:

```python
def inc(x):
    return {'inc': x + 1}

def dec(x):
    return {'dec': x - 1}

def add(x, y):
    return {'add': x + y}


from pyiron_contrib.workflow.node import Node
from pyiron_contrib.workflow.channels import ChannelTemplate

class Inc(Node):
    input_channels = [
        ChannelTemplate("x"),
    ]
    output_channels = [
        ChannelTemplate("inc"),
    ]

inc_node = Inc('inc': inc)


class Dec(Node):
    input_channels = [
        ChannelTemplate("x"),
    ]
    output_channels = [
        ChannelTemplate("dec"),
    ]

dec_node = Dec('dec': dec)


class Add(Node):
    input_channels = [
        ChannelTemplate("x"),
    ]
    output_channels = [
        ChannelTemplate("add"),
    ]

add_node = Add('add': add)
```

And with my suggestion, it looks like this:

```python
def inc(x):
    return x + 1

def dec(x):
    return x - 1

def add(x, y):
    return x + y

from pyiron_contrib.workflow.node import Node

inc_node = Node('inc', inc)
dec_node = Node('dec', dec)
add_node = Node('add', add)
```

For someone who's used to working with Dask, my solution is probably acceptable, but I'm really not sure if they would be willing to do all the effort to write the channels, especially in this case they are just repetitions of names. A bigger problem here is that a usual workflow contains **a lot** of nodes, so this example is certainly not an exaggeration. Moreover, a function can be included into a workflow only if it's included in a node, so the user **must** do all the efforts, if my solution is not available, even if the function is only about adding two variables. Here, I'm not saying that pyiron workflow should be the same as the Dask workflow, but I have difficulties explaining why the user **must** ("must" and not "has the freedom to choose to") write so many more lines in pyiron.

Reply:
All of those functions return a single value. Is there a solution in this format for multiple outputs?

Reply:
Also, in your last example you don't show how to set up the graph. Is it supposed to look like this

```python
add_node.input.x = inc_node.output
```

or 


```python
add_node.x = inc_node
```

Again, I really run up against the fact that the output field has no name. I don't see that being workable for nodes with multiple outputs, and I don't think I want to abandon multiple outputs.

Reply:
In a dask format, I suppose we could write something like 

```python
a, b = returns_multiple("some", "input")
c = another_function(a, b)
```

Which sort of handles the case for text-based interface -- but what about GUIs? We still just have "output0", "output1", etc...

Reply:
I would finally argue that your "current implemenation" example still unfairly bloats lines by going the sub-class route. You can do everything just instantiating a `Node`, and wind up looking a lot closer to your final example:

```python
def inc(x):
    return x + 1

def dec(x):
    return x - 1

def add(x, y):
    return x + y

from pyiron_contrib.workflow.node import Node
from pyiron_contrib.workflow.channels import ChannelTemplate

inc_node = Node(
    'inc', inc, 
    input_channels=[ChannelTemplate('x')], 
    output_channels=[ChannelTemplate('y')]
)
dec_node = Node(
    'dec', dec, 
    input_channels=[ChannelTemplate('x')], 
    output_channels=[ChannelTemplate('y')]
)
add_node = Node(
    'add', add, 
    input_channels=[ChannelTemplate('x'), ChannelTemplate('y')], 
    output_channels=[ChannelTemplate('z')]
)

add_node.input.x = inc_node.output.y
add_node.input.y = inc_node.output.z
```

Again, I'm not saying this is perfect or that I'm closed to changes, but I absolutely want to preserve: named outputs (necessary for a GUI), and multiple outputs (necessary, e.g. for looping functions that need both value output and execution signal output). So far the much more succinct alternatives you're suggesting keep losing these things.

Reply:
> Which sort of handles the case for text-based interface -- but what about GUIs? We still just have "output0", "output1", etc...

But that's up to the user to say whether it's good or not. I don't understand why we should not offer this possibility in the first place.

> The example above is 30 lines (including newlines to keep the instantiation arguments nicely readable), and your dask example is 14. So basically the current implementation pays a factor of 2 in verbosity to gain the features I want. That's a bit of a steep price, but it's not an order of magnitude or anything. Maybe we can get that factor down without losing functionality.

Wait my Dask example is 4 lines, including the import statement. Yours is 18 or 19, depending on the import. And I would say that's a lot more lines.

Reply:
> But that's up to the user to say whether it's good or not. 

Me. I am the user. I have been working with this sort of workflow stuff for the last couple of years in various formats, and I'm telling you that refercing outputs by index (rather than by name) is a huge pain in the ass. I insist that outputs be named.

> I don't understand why we should not offer this possibility in the first place.

I'm not sure what you mean here.

> Wait my Dask example is 4 lines, including the import statement. Yours is 18 or 19, depending on the import. And I would say that's a lot more lines.

Yep, that's probably a better way to count. x5 instead of x2, then.

But it's still a bit beside the point, because the dask example is still missing functionality I want!

Here is a solution I'm not happy with, because I think it is not sufficiently explicit, relies on overloading, and does too much under-the-hood magic, but I could maybe live with it:

```python
def inc(x):
    return x + 1

def dec(x):
    return x - 1

def add(x, y):
    return x + y, x-y

from pyiron_contrib.workflow.node import Node
from pyiron_contrib.workflow.channels import ChannelTemplate

inc_node = Node(function=inc, output_channels=['y'])
dec_node = Node(function=dec, output_channels=['y'])
add_node = Node(
    fuction=add, 
    output_channels=['sum', 'diff'],
    x=inc_node.output.y,
    y=inc_node.output.y
)
```

Where we leverage `inspect` to populate the node name from the function name, and `input_channels` from the function signature. Further we change things so that the node `function` is expected to return a tuple (or single value), and the _order of `output_channels` matters_ and is used to map `function` output to `Node.output` _and_ we allow passing either a full `ChannelTemplate` _or_ just a string (which dynamically creates a `ChannelTemplate` under the hood (again, using `inspect` to populate the `types` if `function` had return type hinting).

We give up some of my "wants" of explicitness and directness and trade them in for dirty evil implicit misdirection, but we do keep all of my "needs" of mandatorily named output, multiple outputs, and giving power users the option to specify much more (e.g. ontological types). In return we get within a 2x of dask's line count -- 8 lines to dask's 4.

I expect nodes to be _used_ much more than they're _written_, and I like our `Node(..., arg=other_node.output.arg)` syntax a lot better than dask's `x = my_delayed_guy(y)`, and this solution also preserves our user syntax nicely.

Reply:
>dirty evil implicit misdirection

(I'm being facetious, btw. It's true I have some reservations, but overall your arguments are pushing this in a better direction and I appreciate your persistence)

Reply:
In fact, we could actually make construction of the channels the default, such that we _always_ construct the channels from the function signature, getting rid of the `types` and `input_channels` arguments altogether. The signature of  the node would then be something like

```python
Node.__init__(
    function: callable,  # Node functionality, returns value or tuple
    output_names: list[str],  # Must have same length as function return
    input_otypes: Optional[dict],  # dict with keys matching function arg names
    output_otypes: Optional[dict],  # dict with keys matching output_names
    **function_kwargs,  # accepts values OR
   # other node channels (in which case it makes a connection, as now)
)
```

We'd still build the channels under the hood, but the users would never instantiate them. Subclassing node would then look something like

```python
class MyNode(Node):
    def __init__(self, **function_kwargs):
        super().__init__(
            self.my_function,
            ['first_out', 'second_out'],
            output_otypes = {
                'first_out': some_onto.SomeThing(),
                'second_out': some_onto.SomeOtherThing(),
            },
            **function_kwargs
       )

    @staticmethod
    def my_function(some_input: int, other_input: str) -> SomeThingClass, OtherThingClass:
        blah blah
        return a, b
```

This way it's not like sometimes you need `input_channels` and sometimes not, nor does `output_channels` change between `list[str]` and `list[ChannelTemplate]`.

Reply:
ok I can live with this.

Reply:
Super. When I get back to this, I'll do it that way and ping you again when it's in its new form.